### PR TITLE
http logging

### PR DESCRIPTION
**Pull request by shashiranjan84**

### CODE REVIEW

File: kong/plugins/httplog/log.lua [hunk 1]
```diff
@@ -0,0 +1,19 @@
+local cjson = require "cjson"
+local http_client = require "kong.tools.http_client"
+
+local APPLICATION_JSON_TYPE = "application/json"
+
+local _M = {}
+
+local function log(premature, conf, message)
+  local response, status, headers = http_client.post(conf.logging_url, cjson.encode(message), { ["content-type"] = APPLICATION_JSON_TYPE })
```
Comment (thibaultCha) [inline kong/plugins/httplog/log.lua]: Please don't put the result in variables if they are unused, this will not pass linting.

File: kong/plugins/httplog/log.lua [hunk 2]
```diff
@@ -0,0 +1,19 @@
+local cjson = require "cjson"
+local http_client = require "kong.tools.http_client"
+
+local APPLICATION_JSON_TYPE = "application/json"
```
Comment (thibaultCha) [inline kong/plugins/httplog/log.lua]: Not really needed, only used once.

File: kong/plugins/httplog/log.lua [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+local _M = {}
+
+local function log(premature, conf, message)
+  local response, status, headers = http_client.post(conf.logging_url, cjson.encode(message), { ["content-type"] = APPLICATION_JSON_TYPE })
+end
+
+function _M.execute(conf)
+  local ok, err = ngx.timer.at(0, log, conf, ngx.ctx.log_message)
```
Comment (thibaultCha) [inline kong/plugins/httplog/log.lua]: Please add `timer.at` to the nginx_stub to allow unit tests to mock it if needed.

File: kong/plugins/httplog/handler.lua [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+local HttpLogHandler = BasePlugin:extend()
+
+function HttpLogHandler:new()
+  HttpLogHandler.super.new(self, "httplog")
+end
+
+function HttpLogHandler:log(conf)
+  HttpLogHandler.super.log(self)
+  log.execute(conf)
```
Comment (thibaultCha) [inline kong/plugins/httplog/handler.lua]: It's really unnecessary to nest this in `log.lua` when your actual function is like 5 lines long. Just keep it simple and put the code in the `handler:log`

**Comment by thibaultCha** [comment]
using the http_client is not a good idea. http_client should only meant to be used for tests and not be part of Kong's public API. It's very basic, not well tested, and not used anywhere else than in this plugin so once this plugin is moved out, there is no guarantee that http_client will not drastically change or even be removed as it is only for testing.

Finally, and more important, it is making synchronous http connections.

Can you describe why you are using `ngx.timer.at`? From the docs:

```
Because timer callbacks run in the background and their running time will not add to any client request's response time, they can easily accumulate in the server and exhaust system resources due to either Lua programming mistakes or just too much client traffic. To prevent extreme consequences like crashing the Nginx server, there are built-in limitations on both the number of "pending timers" and the number of "running timers" in an Nginx worker process. The "pending timers" here mean timers that have not yet been expired and "running timers" are those whose user callbacks are currently running.

The maximal number of pending timers allowed in an Nginx worker is constrolled by the lua_max_pending_timers directive. The maximal number of running timers is controlled by the lua_max_running_timers directive.
```

It seems kinda risky and unstable to use in such a context. I would feel more confident about it if we were opening a connection using the engine's cosocket API (asynchronous) that you would then put back into the nginx pool, which would be a **very** nice performance boost for your logging. If nginx sees a previously opened socket to the configured logging URL in the pool, it will use it instead of opening a new one from scratch.

File: kong/plugins/httplog/log.lua (hunk 1 above)
Comment (thibaultCha) [inline kong/plugins/httplog/log.lua]: Maybe one should be able to configure the HTTP method too, which is another reason not to use the `http_client`

File: kong/plugins/httplog/schema.lua [hunk 5]
```diff
@@ -0,0 +1,3 @@
+return {
+  logging_url = { required = true, type = "string" }
```
Comment (thibaultCha) [inline kong/plugins/httplog/schema.lua]: There are ways of validating the URL, for example with a `reggae` field, like in the DAO's `apis.lua`

File: kong/plugins/httplog/log.lua (hunk 1 above)
Comment (thefosk) [inline kong/plugins/httplog/log.lua]: While I agree on the first comment, I disagree on the second one: any other method other than `POST` is not going to be the best option for this kind of functionality.

Comment (thibaultCha) [inline kong/plugins/httplog/log.lua]: You shouldn't restrict it either, it should probably be a default but you never know. Maybe @ahmadnassri agrees with me on that.

**Comment by thefosk** [comment]
A better solution could be using the [`tcp`](http://wiki.nginx.org/HttpLuaModule#ngx.socket.tcp) object and building the HTTP request manually: it's not hard and the performance will be higher, while leveraging the async I/O of nginx.

**Comment by thibaultCha** [comment]
It's actually exactly what I am suggesting.

File: kong/plugins/httplog/log.lua (hunk 1 above)
Comment (ahmadnassri) [inline kong/plugins/httplog/log.lua]: me loves custom HTTP methods!

yes, `POST` is the defacto choice here, however, there's no reason why a user can dedicate a method for logging (say `LOG`) or use an officially [registered one](http://www.iana.org/assignments/http-methods/http-methods.xhtml) like `REPORT` that's linguistically more appropriate perhaps.

regardless, method naming is contextual to the receiving end and the entity structure, if you think of a log as a *single entity* then sending *entries* should perhaps be done over a `PATCH` request to fit that semantic...

in summary: its best to allow custom methods, and keep a default of `POST`

File: kong/plugins/httplog/log.lua (hunk 2 above)
Comment (ahmadnassri) [inline kong/plugins/httplog/log.lua]: agreed, in the future if there would be format transformers, at that point it would be appropriate to add this back.

**Comment by ahmadnassri** [comment]
are any of those logging plugins sharing parts of logic that can be joined / re-used? seems to me that there might be room for that here ...

**Comment by thibaultCha** [comment]
In the end, logging plugins (http, tcp, up, file) will have a `format` property (`default` (or simple), which will be documented, `ALF` (using the ALF Lua serializer from the API analytics plugin) and possibly `custom`, a function to implement for ex.

It's something we discussed with @thefosk.

**Comment by ahmadnassri** [comment]
nice.

File: kong/plugins/httplog/schema.lua [hunk 6]
```diff
@@ -0,0 +1,6 @@
+return {
+  logging_url = { required = true, type = "string" },
+  method = { require = false, default = "POST", enum = { "POST", "PUT", "PATCH" } },
```
Comment (thibaultCha) [inline kong/plugins/httplog/schema.lua]: Do not enforce an enum, leave it free to be whatever the user wants.

Otherwise, everything looks much better :+1:

Comment (thibaultCha) [inline kong/plugins/httplog/schema.lua]: Also it's `required`, there is a typo here. (And require is false by default so actually no need.

Comment (thefosk) [inline kong/plugins/httplog/schema.lua]: I think the `enum` looks fine and will avoid that 99.9% of the users could write a typo instead (if it's a free form field, we can't prevent typos on the most common methods). For the remaining 0.01% who has a need to use their own HTTP method then we are open to change this.

`GET`, `TRACE`, `CONNECT`, `HEAD` and `OPTIONS` are excluded on purpose because whoever wants to use these methods for logging should be discouraged from doing so.

**Comment by shashiranjan84** [comment]
closing this pull request, will create new one with new branch.
